File: 백준/Gold/2470. 두 용액/두 용액.py
from sys import stdin

if __name__ == "__main__":
    N = int(stdin.readline())
    num_list = []

    for num in stdin.readline().strip().split():    
        num_list.append(int(num))
    num_list = sorted(num_list)
    Left = 0
    Right = len(num_list)-1

    answer = [0,0,abs(num_list[0]+num_list[-1])]
    while(Left<Right):
        if abs(num_list[Left] + num_list[Right]) <= answer[2]:
            answer[0] = num_list[Left]
            answer[1] = num_list[Right]
            answer[2] = abs(num_list[Left] + num_list[Right])

        if answer[2] == 0:
            break

        if num_list[Left]+num_list[Right] > 0:
            Right -=1
        else:
            Left += 1
        
    print(answer[0], answer[1])

File: 백준/Gold/2470. 두 용액/두 용액.cc
#include <iostream>
#include <algorithm>

using namespace std;

int nums[100'001];

int main(void){
    ios::sync_with_stdio(0); // c stream, c++ stream 중 c++ stream만 사용
    cin.tie(0); // 버퍼 비우지 않음

    int n;
    cin >> n;

    for(int i=0; i<n; i++) cin >> nums[i];

    sort(nums, nums+n, less<int>());

    int left=0, right=n-1;
    long long answer = 2'000'000'000;
    int a_left=0, a_right=0;

    while(left<right)
    {
        int sum_nums = nums[left] + nums[right];

        if(abs(answer) > abs(sum_nums))
        {
            answer = sum_nums;
            a_left = left;
            a_right = right;
        }

        if(sum_nums > 0) right--;
        if(sum_nums < 0) left++;
        if(sum_nums == 0) break;
    }

    cout << nums[a_left] << " " << nums[a_right];
    return 0;
}